stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -335,15 +336,25 @@ private void drainQuietly() {
   private final class QueuingListener extends ClientCall.Listener<RespT> {
     @Override
     public void onMessage(RespT value) {
-      Preconditions.checkState(closedStatus == null, "ClientCall already closed");
+      Preconditions.checkState(closeState.get() == null, "ClientCall already closed");
       buffer.add(value);
     }
 
     @Override
     public void onClose(Status status, Metadata trailers) {
-      Preconditions.checkState(closedStatus == null, "ClientCall already closed");
-      closedStatus = status;
+      CloseState newCloseState = new CloseState(status, trailers);
+      boolean wasSet = closeState.compareAndSet(null, newCloseState);
```

Comment:
This is the only place assigning the value, so the code was fine with a plain volatile. It only runs on one thread. The volatile is because it is read by other threads, but it is read-only.